main/test/ca/uwaterloo/flix/library/TestRedBlackTree.flix

```diff
@@ -224,37 +224,97 @@ namespace TestRedBlackTree {
-    def mapWithKey05(): Bool =
+    def seqMapWithKey05(): Bool =
         let Map(t) = Map#{2 -> -1, 5 -> 15, 11 -> -9, 8 -> 8};
         Map(RedBlackTree.seqMapWithKey((k, v) -> k + v, t)) == Map#{2 -> 1, 5 -> 20, 11 -> 2, 8 -> 16}
 
-    // TODO: Observe effects are sequential
+
+    /////////////////////////////////////////////////////////////////////////////
+    // seqMapWithKey fusion                                                    //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def seqMapWithKeyFusion01(): Bool & Impure =
+        let l = ref Nil;
+        let Map(t) = Map#{1 -> 1, 2 -> 2, 3 -> 3};
+        t |>
+        RedBlackTree.seqMapWithKey((_, v) -> { l := "a" :: deref l; v }) |>
+        RedBlackTree.seqMapWithKey((_, v) -> { l := "b" :: deref l; v });
+        List.reverse(deref l) == "a" :: "a" :: "a" :: "b" :: "b" :: "b" :: Nil
+
+    @test
+    def seqMapWithKeyFusion02(): Bool & Impure =
+        let l = ref Nil;
+        let Map(t) = Map#{1 -> 1, 2 -> 2, 3 -> 3};
+        t |>
+        RedBlackTree.seqMapWithKey((_, v) -> { l := "a" :: deref l; v } as & Pure) |>
+        RedBlackTree.seqMapWithKey((_, v) -> { l := "b" :: deref l; v } as & Pure);
+        List.reverse(deref l) == "a" :: "a" :: "a" :: "b" :: "b" :: "b" :: Nil
+
 
     /////////////////////////////////////////////////////////////////////////////
     // parMapWithKey                                                           //
     /////////////////////////////////////////////////////////////////////////////
 
-    // TODO: Observe effects are parallel
+    @test
+    def parMapWithKey01(): Bool =
+        let Map(t) = Map#{}: Map[Unit, Int32];
+        Map(RedBlackTree.parMapWithKey((k, v) -> k + v, t)) == Map#{}
+
+    @test
+    def parMapWithKey02(): Bool =
+        let Map(t) = Map#{1 -> 4};
+        Map(RedBlackTree.parMapWithKey((k, v) -> k + v, t)) == Map#{1 -> 5}
+
+    @test
+    def parMapWithKey03(): Bool =
+        let Map(t) = Map#{2 -> -1, 0 -> 0};
+        Map(RedBlackTree.parMapWithKey((k, v) -> k + v, t)) == Map#{2 -> 1, 0 -> 0}
+
+    @test
+    def parMapWithKey04(): Bool =
+        let Map(t) = Map#{2 -> -1, 5 -> 15, 11 -> -9};
+        Map(RedBlackTree.parMapWithKey((k, v) -> k + v, t)) == Map#{2 -> 1, 5 -> 20, 11 -> 2}
+
+    @test
+    def parMapWithKey05(): Bool =
+        let Map(t) = Map#{2 -> -1, 5 -> 15, 11 -> -9, 8 -> 8};
+        Map(RedBlackTree.parMapWithKey((k, v) -> k + v, t)) == Map#{2 -> 1, 5 -> 20, 11 -> 2, 8 -> 16}
+
+
+    /////////////////////////////////////////////////////////////////////////////
+    // parMapWithKey fusion                                                    //
+    /////////////////////////////////////////////////////////////////////////////
+
+    // This test tries to force a lost update
+    // but there **could** be a run where all updates are accounted for
+    // so how do we ensure the parallelism is observed deterministically?
```

Comment:
So I think thats its only for the *seq* version that you should test the order is as expected.

Reply:
Yeah, this simply tests an integer value

Reply:
or wait should we not observe that parmap has parallel side effects?

Reply:
> 

No. Thats okay.

Reply:
What I want is that `map` and `mapWithKey` behave correctly when called with an impure function (i.e. seeing the tree in order).

Reply:
I think we should just remove this test.

What we should have, I now think, are tests that call `RedBlackTree.map` (not seq/par).

- Most tests should just verify that the results are correct. (These will implicitly use parallelism as long as we ensure the trees are of some size, e.g. 2k elements).
- A few tests should be in the style of `seqMapWithKeyFusion01` but they should just call `RedBlackTree.map`.

Reply:
> A few tests should be in the style of seqMapWithKeyFusion01 but they should just call RedBlackTree.map.

And not `RedBlackTree.mapWithKey` instead of `RedBlackTree.map`?

Reply:
I am looking for tests that test the whole pipeline. I believe that map calls mapWithKey so that would be fine.

Reply:
I am basically looking for 5-8 tests on trees of various sizes (but at least handful larger than the threshold) that call RBTree.map with a pure function and compare the results.

And 5-8 tests that call RBTree.map with an impure function and records the order of effects by storing something into a mutable data structure etc.

Reply:
Does it make sense?

Reply:
Yes I think so. I'll add it and ping you again :+1: